### PR TITLE
feat(gemma4): paged KV serving across both attention families

### LINKED ISSUE #884: gemma4: every step recomputes the whole prompt
## Description

The 12B tower forward is the correctness building block: it embeds the prompt, runs all 48 layers, and projects logits, with no cache. Generating a token that way costs a full prompt recomputation, so nothing above it can serve.

Gemma 4 cannot reuse the single-pool KV machinery the full-attention models ride. Its layers alternate two families with different geometry and different lifetimes: local layers are sliding-window attention at head_dim 256 with 8 KV heads, global layers are full attention at head_dim 512 with 1 KV head. One pool with one page geometry cannot hold both, and a request holds pages in both at once — so admission has to be atomic across two pools or a half-admitted request leaks the granted half.

## Proposed Plan

- One pool per family, each with its own page geometry; a request holds one `KvState` per family plus its absolute frontier.
- Atomic dual-pool admission: reserve the exact page delta from both pools, commit only when both reservations hold, and let a refused side roll the other back by drop.
- A step plan as the executor boundary — the layer forwards take the plan plus the server-owned pools, never a request's state, so window crossing later changes the plan builder rather than the layer forwards.
- Prefill and single-token decode steps over the read-only attention entries the pool-write prep feeds; decode runs the prefill entry at seq_len 1. The greedy loop that drives the reference gate is test-side — production sampling belongs to the serving frontend.
- Fail-closed at the sliding window: cache slots and absolute positions coincide only below it, and this slice refuses to cross rather than silently conflating the two coordinate systems.

## Acceptance Criteria

- Greedy continuation matches HF `generate()` token for token on three prompts.
- A forced partial admission leaves both pools at their pre-request occupancy.
- Pages return to both pools when a request drops, and the next request gets them.
- The KV path agrees with the no-KV tower forward on the same prompt — argmax for argmax, within a calibrated logit tolerance.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #884.

Serving the 12B tower needs two KV families, not one: local layers are sliding-window attention at head_dim 256 with 8 KV heads, global layers are full attention at head_dim 512 with 1 KV head. One pool with one page geometry cannot hold both, and a request holds pages in both at once. This adds a pool per family, a per-request state in each, and the step machinery over them — prefill, greedy decode, and the admission that keeps the two pools consistent. The pool-write prep (#881) and the windowed attention wrapper (#883) it rides on are already in.

- **Admission is atomic across both pools.** The page count is the exact frontier account — a ceiling over the post-step KV length per family, minus what the request already holds. Both reservations are taken before either is committed; a refused side drops the other, and both pools sit at their pre-request occupancy. A half-admitted request would otherwise strand the granted family's pages for the request's lifetime.
- **The layer forwards take the plan and the pools, never a request's state.** Both families' attention plans are built over the same token span; the buffer and page geometry come from the pools the server owns, and everything request-scoped reaches the kernels as plan metadata. Request state is touched only at the step boundary, so moving the resident window later changes the plan builder, not the layer forwards.
- **Decode runs the prefill entry at seq_len 1.** Correct first; the dedicated windowed decode form is a measurement-gated decision for the batching stage. The greedy loop that drives the reference gate lives with the gate, not in the serving module — production sampling is the frontend's.
- **Nothing changes state before the step is admissible.** The tokens are checked against the embedding's rows (the kernel indexes it without validating), both families must sit at the same frontier and hold exactly the pages the frontier accounts for, and the window and rope bounds are checked — all before either family advances, because a `KvState` only ever moves forward.
- **Page ids only mean something against the pool that issued them.** A request's state is checked to belong to this server's pools before admission commits or the executor indexes a buffer with its pages.
- **Model identity is structural, not checked.** The server holds the weights its pools, rope tables and layer numbering were built from, so a step cannot be run against a different checkpoint — pages written under one are meaningless under another, and a same-shape mismatch would pass every geometry check. Device identity is pinned on both sides: construction refuses weights that live on a different GPU than the one the pools are allocated on, and a step refuses a context from another device.
- **Fail-closed at the window.** Cache slots and absolute positions coincide only below the sliding window, and this path refuses to cross.

## Test Env

Single GPU (sm_89, x86_64), CUDA 12.9.

## Verification

- `cargo fmt --all -- --check`; `cargo build`/`cargo clippy --release -p pegainfer-gemma4 --features gemma4 --all-targets -- -D warnings`; the same pair featureless — all clean. 26 library tests pass.
- Light GPU test: atomic dual-pool admission rollback — a forced partial failure leaves both pools at their pre-request occupancy.
- KV A/B (paged serving vs the no-KV oracle forward, same weights and tokens): prefill argmax equal at every one of 9 positions, max |dlogit| 1.3125; four decode steps fed the oracle's continuation, argmax equal, max |dlogit| 0.90625/0.953125/0.875/0.625 — all under the calibrated 2.0 gate.
- Greedy generation matches HF `generate()` token for token on three prompts, 50 tokens each. The dumper refuses anything that is not reachable rather than recording it: sdpa and eager must agree on every token, a replay must reproduce the run, and each case is truncated at the first step whose top1-top2 margin falls to the gate. The three admitted cases measured minimum margins of 2.69/5.62/6.81 at dump time — over 26 candidate prompts, exactly three stayed decisive for 50 straight steps, because near-tie steps flip between independent implementations even where HF agrees with itself (flips observed up to margin 1.69). The fixture records the gate and the agreement claim, not the per-step margins themselves.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: pegainfer-gemma4/src/forward.rs [hunk 1]
```diff
@@ -29,6 +29,29 @@ pub(crate) fn embed_scale_bf16(hidden_size: usize) -> f32 {
...
+    anyhow::ensure!(
+        weights.embed_tokens.cols == hidden_size,
+        "embedding width {} != config hidden_size {hidden_size}",
+        weights.embed_tokens.cols
+    );
+    for (position, &token) in tokens.iter().enumerate() {
+        anyhow::ensure!(
+            (token as usize) < vocab_size,
+            "token {token} at position {position} is outside the embedding's {vocab_size} rows"
+        );
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/forward.rs:56]: Good catch

File: pegainfer-gemma4/src/serve.rs [hunk 2]
```diff
@@ -0,0 +1,509 @@
+//! KV-backed serving forward: two KV families, short-context prefill and
+//! decode, and atomic dual-pool admission.
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/serve.rs:2]: The durable KV contract is already owned by #809, which adds docs/models/gemma4/kv-cache.md and its index entry.

**Comment by FeathBow** [comment]
@codex review please

File: pegainfer-gemma4/src/serve.rs [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+        span: LogitsSpan,
+    ) -> Result<HiddenStates> {
+        let seq_len = tokens.len();
+        anyhow::ensure!(seq_len > 0, "step needs at least one token");
+        let weights = &self.weights;
+        anyhow::ensure!(
+            ctx.device_ordinal == self.local_pool.buffer().ordinal(),
+            "this context is on device {} but the pools live on device {}",
+            ctx.device_ordinal,
+            self.local_pool.buffer().ordinal()
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/serve.rs]: Good catch. Device equality does not establish stream ordering here because event tracking is disabled. I’ll bind the serving path to the construction stream so pool/RoPE initialization and successive KV writes remain ordered without adding per-step synchronization.

**Comment by FeathBow** [comment]
@codex check again please

File: pegainfer-gemma4/src/serve.rs [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        let weights = &self.weights;
+        anyhow::ensure!(
+            std::sync::Arc::ptr_eq(&ctx.stream, self.local_pool.buffer().stream()),
+            "step must use the DeviceContext stream that constructed this GemmaServe"
+        );
+        validate_tokens(weights, self.local_geom.hidden_size, tokens)?;
+        let start_pos = kv.local.seq_len();
+        self.check_step_bounds(kv, start_pos + seq_len)?;
+        kv.local.advance(seq_len);
+        kv.global.advance(seq_len);
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/serve.rs]: Great. The current plan construction depends on a prematurely advanced frontier, and an Err leaves reusable-looking state that may describe unwritten KV.

**Comment by FeathBow** [comment]
@codex review again please
